File: 2021/csharp/day15/Program.cs
var rawCosts = File.ReadAllLines("input.txt")
    .Select(t => t.Select(t => t - '0').ToArray())
    .ToArray();

int Traverse (Graph graph, (int x, int y) source, (int x, int y) destination)
{
    var vertices = Enumerable.Range(0, graph.Length).SelectMany(v => Enumerable.Range(0, graph.Length).Select(y => (v, y))).ToHashSet();
    var distances = new Dictionary<(int x, int y), int>();
    var priorityQueue = new PriorityQueue<(int x, int y), int>();

    distances[(0, 0)] = 0;
    priorityQueue.Enqueue((0, 0), 0);
    while (priorityQueue.TryDequeue(out (int x, int y) currentNode, out int currentCost))
    {
        if (!vertices.Remove(currentNode))
            continue;

        var neighbours = new[] { (1, 0), (-1, 0), (0, 1), (0, -1) }
            .Select(t => (t.Item1 + currentNode.x, t.Item2 + currentNode.y))
            .Where(vertices.Contains)
            .ToArray();

        foreach (var neighbour in neighbours)
        {
            var newCost = currentCost + graph[neighbour];
            if (newCost < distances.GetValueOrDefault (neighbour, int.MaxValue))
            {
                distances[neighbour] = newCost;
                priorityQueue.Enqueue(neighbour, newCost);
            }
        }
    }
    return distances[destination];
}

var graphQ1 = new Graph(rawCosts, 1);
var graphQ2 = new Graph(rawCosts, 5);

Console.WriteLine($"Q1 cost: {Traverse(graphQ1, (0, 0), (graphQ1.Length - 1, graphQ1.Length - 1))}");
Console.WriteLine($"Q2 cost: {Traverse(graphQ2, (0, 0), (graphQ2.Length - 1, graphQ2.Length - 1))}");

public readonly struct Graph
{
    readonly int[][] graph;
    readonly int tiles;

    public int Length => graph.Length * tiles;

    public int this[(int x, int y) index]
    {
        get
        {
            int additionX = index.x / graph.Length;
            int additionY = index.y / graph.Length;

            var cost = graph[index.x % graph.Length][index.y % graph.Length] + (additionX + additionY);
            while (cost > 9)
                cost -= 9;
            return cost;
        }
    }

    public Graph(int[][] graph, int tiles)
        => (this.graph, this.tiles) = (graph, tiles);
}

File: 2021/fsharp/day15/Program.fs
open System.IO
open System.Collections.Generic

let parseInput(filePath) =
    let parseChar c = (c |> int) - ('0' |> int)

    File.ReadAllLines (filePath)
    |> Array.map(fun value -> (value.ToCharArray () |> Array.map parseChar))

let traverse(accessor, length) =
    let mutable vertices =  [|0..length - 1|]
                            |> Array.collect(fun x -> ([|0..length - 1|] |> Array.map (fun y -> (x, y))))
                            |> Set.ofArray

    let mutable distances = Map.empty<(int*int), int>
    let pq = new PriorityQueue<int*int, int> ();
    
    pq.Enqueue((0, 0), 0);
    distances <- distances.Add((0, 0), 0)
    while (pq.Count > 0) do 
        let _, currentNode, currentCost = pq.TryDequeue();
        if vertices.Contains (currentNode) then
            vertices <- vertices.Remove (currentNode)

            [|(0,  1) ; (0, -1); (1, 0) ; (-1, 0) |]
                |> Array.map (fun (x, y) -> (x + (currentNode |> fst), y + (currentNode |> snd)))
                |> Array.filter (vertices.Contains)
                |> Array.map(fun t -> (t, currentCost + accessor((t |> fst),(t |> snd))))
                |> Array.filter(fun (node, cost) -> cost < distances.GetValueOrDefault (node, System.Int32.MaxValue))
                |> Array.iter (fun (node, cost) -> (distances <- distances.Add (node, cost) ;
                                                    pq.Enqueue(node, cost)))

    distances[(length - 1, length - 1)]
    

let execute =
    let grid_extender(grid:int[][])(x:int, y:int) =
        (grid[x % grid.Length][y % grid.Length] + x / grid.Length + y / grid.Length - 1) % 9 + 1

    "input.txt"
    |> parseInput
    |> fun input -> [|("Q1", input, input.Length) ; ("Q2", input, input.Length * 5) |]
    |> Array.map (fun (q, grid, length) -> (q, traverse(grid_extender grid, length)))
    |> Array.map (fun (q, danger) -> printfn "%s danger: %d" q danger)

execute |> ignore